files/zh-cn/web/javascript/reference/statements/for...in/index.md

```diff
@@ -1,107 +1,234 @@
+
+根据现代 ECMAScript 规范的定义，遍历的顺序是一致且可预测的。在原型链的每个组件中，所有非负整数键（可以作为数组索引）将首先按值升序遍历，然后是其他字符串键按属性创建的先后顺序升序遍历。
+
+在 `for...in` 循环中，`variable` 部分可以接受任何在 `=` 运算符之前的内容。只要在循环体内部不重新赋值（可以在迭代之间更改，因为它们是两个独立的变量），你可以使用 {{jsxref("Statements/const", "const")}} 来声明变量。否则，你可以使用 {{jsxref("Statements/let", "let")}}。你可以使用[解构赋值](/zh-CN/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment)来分配多个局部变量，或者使用属性访问器（例如 `for (x.y in iterable)`）将值分配给对象属性。
+
+一种[遗留的语法](/zh-CN/docs/Web/JavaScript/Reference/Deprecated_and_obsolete_features#语句)允许在 `for...in` 循环中的 `var` 声明的循环变量具有初始值。在严格模式下，这会抛出[语法错误](/zh-CN/docs/Web/JavaScript/Reference/Errors/Invalid_for-in_initializer)，而在非严格模式下则会被忽略。
+
+### 删除、添加或修改的属性
+
+`for...in` 按以下方式访问属性键：
+
+1. 它首先获取当前对象的所有自有的字符串键，其方式与 {{jsxref("Object.getOwnPropertyNames()")}} 非常相似。
+2. 对于每一个键，如果没有访问过具有相同值的字符串，则[获取属性描述符](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Reflect/getOwnPropertyDescriptor)并只访问可枚举的属性。但是，即使该属性不可枚举，也会标记为已访问。
+3. 然后，当前对象被替换为其原型，并重复上述过程。
+
+这意味着：
+
+- 任何在迭代过程中被添加到当前访问对象中的属性都不会被访问到，因为所有当前对象的自有属性已经被保存了。
+- 如果原型链中多个对象具有相同名称的属性，则只会访问第一个，并且只有它是可枚举时才会被访问。如果它是不可枚举的，则不会访问具有相同名称的任何其他属性，即使它们也是可枚举的。
+
+一般来说，最好不要在迭代过程中添加、修改或删除对象属性，除非是正在访问的属性。规范明确允许实现可以在以下情况下不遵循上述算法：
+
+- 在迭代过程中修改对象的原型链。
+- 在迭代过程中从对象或其原型链中删除属性。
+- 在迭代过程中向对象原型链中添加属性。
+- 在迭代过程中更改属性的可枚举性。
+
+在这些情况下，实现可能与你期望的行为不同，甚至可能与其他实现不同。
+
+### 数组迭代和 for...in
 
-### 数组迭代和 `for...in`
+数组索引只是具有整数名称的可枚举属性，除此之外与一般对象属性完全相同。`for...in` 循环将在遍历其他键之前遍历所有整数键，并且按照严格递增的顺序进行，使得 `for...in` 的行为接近正常的数组迭代。然而，`for...in` 循环会返回所有可枚举属性，包括那些具有非整数名称和被继承的属性。与 `for...of` 不同，`for...in` 使用属性枚举而不是数组的迭代器。在[稀疏数组](/zh-CN/docs/Web/JavaScript/Guide/Indexed_collections#稀疏数组)中，`for...of` 会访问空槽，但 `for...in` 不会访问空槽。
 
-> **备注：** `for...in`不应该用于迭代一个关注索引顺序的 {{jsxref("Array")}}。
+最好使用带有数值索引的 {{jsxref("Statements/for", "for")}} 循环、{{jsxref("Array.prototype.forEach()")}} 或 {{jsxref("Statements/for...of", "for...of")}} 循环，因为它们会将索引作为数字而不是字符串返回，并且还会避免非索引属性的干扰。
 
 ### 仅迭代自身的属性
 
-如果你只要考虑对象本身的属性，而不是它的原型，那么使用 {{jsxref("Object.getOwnPropertyNames", "getOwnPropertyNames()")}} 或执行 {{jsxref("Object.prototype.hasOwnProperty", "hasOwnProperty()")}} 来确定某属性是否是对象本身的属性（也能使用{{jsxref("Object.prototype.propertyIsEnumerable", "propertyIsEnumerable")}}）。或者，如果你知道不会有任何外部代码干扰，你可以使用检查方法扩展内置原型。
+如果你只想迭代对象本身的属性，而不是它的原型，你可以使用以下技术之一：
 
-## 为什么用 for ... in?
+- {{jsxref("Object.keys()")}}
+- {{jsxref("Object.getOwnPropertyNames()")}}
 
-`for ... in`是为遍历对象属性而构建的，不建议与数组一起使用，数组可以用`Array.prototype.forEach()`和`for ... of`，那么`for ... in`的到底有什么用呢？
+`Object.keys` 会返回一个包含所有可枚举的自有字符串属性的数组，而 `Object.getOwnPropertyNames` 则会包含所有属性，包括不可枚举的。
 
-它最常用的地方应该是用于调试，可以更方便的去检查对象属性（通过输出到控制台或其他方式）。尽管对于处理存储数据，数组更实用些，但是你在处理有`key-value`数据（比如属性用作“键”），需要检查其中的任何键是否为某值的情况时，还是推荐用`for ... in`。
+许多 JavaScript 风格指南和代码检查工具建议避免使用 `for...in` 循环，因为它会遍历整个原型链，这很少是我们想要的，并且可能会与更常用的 `for...of` 循环混淆。`for...in` 循环最常用于调试目的，它是一种简单的方式来检查对象的属性（通过向控制台输出或其他方式）。在对象被用作临时键值对的情况下，`for...in` 循环可以检查这些键中是否有特定的值。
```

Comment:
> For the last sentence, I guess it gives the usage of checking the value (instead of the key) in the key-value pair, for checking the key we can use a more suitable tool - `set`, or `key in obj` syntax

I agree with you, maybe my translation is ambiguity. I will fix it.